src/server/tiering/decoders.h

```diff
@@ -0,0 +1,61 @@
+// Copyright 2025, DragonflyDB authors.  All rights reserved.
+// See LICENSE for licensing terms.
+//
+
+#pragma once
+
+#include <memory>
+#include <optional>
+#include <string>
+#include <string_view>
+
+#include "core/compact_object.h"
+#include "core/string_or_view.h"
+
+namespace dfly::tiering {
+
+// Decodes serialized value and provides it to callbacks.
+// Acts as generic interface to callback driver (OpManager)
+struct Decoder {
+  virtual ~Decoder() = default;
+
+  // Poor man's type-erasure copy
+  virtual std::unique_ptr<Decoder> Clone() const = 0;
```

Comment:
1. Either this, so the caller can pass a decoder by value, or the caller has to pass a function creating an encoder
2. I don't want to use unique_ptrs later, I want to initialize in place with proper align to avoid allocations